### PR TITLE
Add the ability to preopen dirs during WASI initialization

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -39,8 +40,8 @@ const DEFAULT_WASM_MULTI_MEMORY: bool = true;
 ///
 /// * Reference types are not supported yet. This is tricky because it would
 ///   allow the Wasm module to mutate tables, and we would need to be able to
-///   diff the initial table state with the new table state, but funcrefs and
-///   externrefs aren't comparable in the Wasm spec, which makes diffing
+///   snapshot the initial table state with the new table state, but funcrefs
```
Comment (cfallin) [inline src/lib.rs:43]: just "snapshot the new table state" maybe? ("snapshot A with B" unclear to me)

File: src/lib.rs [hunk 2]
```diff
@@ -39,8 +40,8 @@ const DEFAULT_WASM_MULTI_MEMORY: bool = true;
...
+///   and externrefs aren't comparable in the Wasm spec, which makes snapshoting
```
Comment (cfallin) [inline src/lib.rs:44]: s/snapshoting/snapshotting/

File: src/lib.rs [hunk 3]
```diff
@@ -630,13 +659,13 @@ impl Wizer {
         Ok(instance)
     }
 
-    /// Diff the given instance's globals, memories, and tables from the Wasm
-    /// defaults.
+    /// Snapshot the given instance's globals, memories, and tables from the
+    /// Wasm defaults.
```
Comment (cfallin) [inline src/lib.rs:663]: "... from the Wasm defaults" no longer applicable?

File: src/lib.rs [hunk 4]
```diff
@@ -970,8 +999,9 @@ impl Wizer {
     }
 }
 
-/// A "diff" of Wasm state from its default value after having been initialized.
-struct Diff<'a> {
+/// A "snapshot" of Wasm state from its default value after having been
```
Comment (cfallin) [inline src/lib.rs:1002]: s/from its default value//?

**Review by cfallin — APPROVED** [review]
Looks good!

Just a few wording nits below in comments etc -- I agree that "snapshot" is clearer than "diff" but "snapshot A with B" / "snapshot A from B"-type phrases I think are no longer applicable (a snapshot is a thing that is conceptually independent from a starting state, at least in my mind).

File: src/lib.rs (hunk 4 above)
Comment (fitzgen) [reply src/lib.rs:1002]: Gah this is what I get for blind replacement! Thanks!

**Pull request merged**